src/app/api/authznet/callbacks/route.ts

```diff
@@ -0,0 +1,237 @@
+            "Could not lookup purchase by PO Number '%s': %O",
+            payload.merchantReferenceId,
+            purchase,
+        );
+        return;
+    }
+
+    const { id: purchaseId, username } = purchase;
+    const transactionResponse = {
+        transId: payload.id,
+        transDate: eventDate,
+        avsResultCode: payload.avsResponse,
+        ...payload,
+    };
+    addTransactionResponse(purchaseId, { transactionResponse });
+
+    if (payload.responseCode !== TransactionResponseCodeEnum.APPROVED) {
+        logger.info("payment not approved %o", {
+            purchaseId,
+            username,
+            payload,
+        });
+        return;
+    }
+
+    if (
+        ![
+            "net.authorize.payment.authcapture.created",
+            "net.authorize.payment.fraud.approved",
+        ].includes(eventType)
+    ) {
+        logger.warn("APPROVED payment for unexpected eventType %o", {
+            purchaseId,
+            username,
+            eventType,
+            payload,
+        });
+    }
+
+    const orderDetails = {
+        ...purchase,
+        transactionResponses: [
+            transactionResponse,
+            ...purchase.transactionResponses,
+        ],
+    };
+
+    // The payment was successful, so update the user's subscription and addons.
+    updateSubscription(username, orderDetails);
+}
+
+async function updateSubscription(
+    username: string,
+    orderDetails: OrderDetails,
+) {
+    const subscription = orderDetails.lineItems?.find(
+        (item) => item.itemId === LineItemIDEnum.SUBSCRIPTION,
+    );
+    const addons = orderDetails.lineItems?.filter(
+        (item) => item.itemId === LineItemIDEnum.ADDON,
+    );
+
+    if (!(subscription || addons)) {
+        serviceAccountEmailAdmin(username, orderDetails);
+        return;
+    }
+
+    const currentSubscriptionResponse =
+        await serviceAccountFetchSubscription(username);
+
+    const currentSubscriptions: UserSubscriptionListing =
+        await currentSubscriptionResponse.json();
+
+    const currentSubscription =
+        currentSubscriptionResponse.ok &&
+        currentSubscriptions?.result?.total > 0 &&
+        currentSubscriptions.result.subscriptions[0];
+
+    if (!currentSubscriptionResponse.ok || !currentSubscription) {
+        logger.error(
+            "Could not fetch user's current subscription: %o",
+            currentSubscriptions,
+        );
+        serviceAccountEmailAdmin(username, orderDetails);
+        return;
+    }
+
+    let addonSubscriptionId = currentSubscription.id;
+
+    let subscriptionUpdateResult;
+    if (subscription) {
+        subscriptionUpdateResult = await serviceAccountUpdateSubscription(
+            currentSubscription,
+            subscription.name,
+            subscription.quantity,
+        );
+
+        if (currentSubscription.plan.name === constants.PLAN_NAME_BASIC) {
+            addonSubscriptionId = subscriptionUpdateResult?.result?.id;
+        }
```

Comment:
Will this fail if a user is renewing a paid subscription (within 30 days of the current subscription's expiration date) and adds add-ons to the subscription at the same time?